src/Perl6/Optimizer.nqp

```diff
@@ -1578,7 +1578,9 @@ my class SmartmatchOptimizer {
 
             return nqp::null() if $lhs.value.HOW.archetypes.generic;
 
-            my $matches := nqp::istype($lhs.value, $rhs.value);
+            # Wrap into try because for if there a user-defined `where`-block involved into typematching it might throw.
+            # Consider this case a failed typematch and proceed further.
+            my $matches := try { nqp::istype($lhs.value, $rhs.value) };
```

Comment:
You shouldn't need a block for either of hese `try`s

Reply:
You're right. I've been overcautious here.